src/charm.py

```diff
@@ -484,14 +559,42 @@ def replace_images(self, resources, registry):
                             [registry] + container["image"].split("/")[-2:]
                         )
 
-    def replace_node_selector(self, resource):
-        label = self.model.config["control-plane-node-label"]
-        label_key, label_value = label.split("=")
+    def replace_node_selector(self, resource, new_label, key_to_replace):
+        label_key, label_value = new_label.split("=")
 
         node_selector = resource["spec"]["template"]["spec"]["nodeSelector"]
-        del node_selector["kube-ovn/role"]
+        del node_selector[key_to_replace]
         node_selector[label_key] = label_value
 
+    def replace_name(self, resource, new_name):
+        resource["metadata"]["name"] = new_name
+
+    def label_bgp_nodes(self, node_selector):
+        label_key, label_value = node_selector.split("=")
+        nodes = json.loads(self.kubectl("get", "nodes", "-o", "json"))["items"]
+        for node in nodes:
+            if node["metadata"]["labels"].get(label_key) == label_value:
+                log.info(
+                    f"Labeling node {node['metadata']['name']} with ovn.kubernetes.io/bgp=true"
+                )
+                self.kubectl(
+                    "label",
+                    "nodes",
+                    node["metadata"]["name"],
+                    "ovn.kubernetes.io/bgp=true",
+                )
+
+    def unlabel_bgp_nodes(self):
+        nodes = json.loads(self.kubectl("get", "nodes", "-o", "json"))["items"]
+        for node in nodes:
+            if node["metadata"]["labels"].get("ovn.kubernetes.io/bgp"):
+                log.info(
+                    f"Removing ovn.kubernetes.io/bgp label from node {node['metadata']['name']}"
```

Comment:
What if I configured my speakers with `node-selector: ovn.kubernetes.io/bgp=true`? I might be tempted to use that label after reading upstream kube-ovn documentation. I think it could lead to some quite surprising behavior - the kube-ovn charm would delete the label, and then there'd be no nodes matching the node selector, right?

I think if we are automatically managing the `ovn.kubernetes.io/bgp` label then we need to blacklist it from the speaker node-selector field and inform the user via Blocked status that they need to pick a different node-selector.

Backing up a bit, do we need the `ovn.kubernetes.io/bgp` label? Is it used by kube-ovn for anything other than node selection? Can we dodge this corner case by skipping the automatic labeling?

Reply:
The node label was for scheduling only, so I removed the logic for labeling and unlabeling the nodes. The node selector passed in to config is enough.